Fix default CTEManager so it can use from_queryset corectly

pgammans opened the pull request:
This commit changes the way that django-cte manager select the queryset to use.

The issues is that the current code forces a new manager to be written for each queryset this is fine if the code is part of an application.  If the code is part of a module then this isn't always controllable.

```
class DataQuerySet(CTEQuerySet):
    def is_default_value(self):
        return self.filter(value = 5)

class DataModel(Model):
    objects = CTEManager.from_queryset(DataQuerySet)()
    value = IntegerField(default=5)

results = DataModel.objects.is_default_value()
```

millerdev commented on django_cte/cte.py:
```diff
@@ -125,7 +125,20 @@ class CTEManager(Manager):
+        if isinstance(qs, CTEQuerySet):
+            return qs
+        else:
+            assert type(qs) == QuerySet, \
+                "models with CTE support need to use a CTEQuerySet"
```
What do you think of changing this to raise `TypeError` instead of `AssertionError`? It seems like a person using django-cte could hit this error if they configured or used the library incorrectly. Assertion errors should only ever be triggered by bugs in the library code itself.

millerdev commented on django_cte/cte.py:
```diff
@@ -125,7 +125,20 @@ class CTEManager(Manager):
+
+    @classmethod
+    def from_queryset(cls, queryset_class, class_name=None):
+        assert issubclass(queryset_class, CTEQuerySet), \
+            "models with CTE support need to use a CTEQuerySet"
```
Same here: `TypeError` rather than `AssertionError`

pgammans replied:
I'm happy with this, I now raise TypeError.
Django seams to use both on model class creation which is where this will happen.

millerdev commented on tests/test_cte.py:
```diff
@@ -14,7 +14,10 @@
 from django_cte import With
 
-from .models import Order, Region
+from .models import (
+    Order,
+    Region,
+)
```
nit: wrapping seems unnecessary here. The line is not too long without it.

millerdev commented on tests/models.py:
```diff
@@ -20,25 +21,75 @@ def lt40(self):
+
 
+class LT40Manager(CTEManager):
+    # This is a semi broke manager but allows testing of the documented
+    # use of get_queryset. It is broken by ignoring the queryset passed
+    # to it via the static method from_queryset
```
What do you think of removing this broken thing and fixing (or removing) the test so it does not rely on it? If it requires a backward-incompatible change we can push a major version and document it, along with a suggestion for how to fix broken code, in the change log.

pgammans replied:
I had added the check to see if i introduced any backward-incompatible changes.

I think we should document the 'new' way and simpler way but the old will still work but didn't want to change the manager and remove test for the documented manger setup in the same commit.

millerdev commented on tests/models.py:
```diff
@@ -20,25 +21,75 @@ def lt40(self):
+    This was the original documented manager setup a custom
+    manager and queryset.
+
+    This is fine if the model isn't going to ge in another
```
```suggestion
    This is fine if the model isn't going to be in another
```

pgammans commented:
I have reworked the default CTEmanager again and dropped the custom get_queryset class and all test still pass.

You can do, but not recommended as skip checks
`
class DataModel(model):
    objects = Manager.from_queryset(CTEQuerySet)
    name = TextField()
`
now as there is no special code in the manager other than a checking the queryset class

millerdev commented on tests/test_manager.py:
```diff
@@ -0,0 +1,110 @@
+            "This should be a CTEQuerySet if we want this to work"
+
+        with self.assertRaises(TypeError):
+            class BrokenModel(Order):
+                class Meta:
+                    proxy = True
+                objects = CTEManager.from_queryset(BrokenQuerySet)()
```
The class and manager instantiation seem unnecessary
```suggestion
            CTEManager.from_queryset(BrokenQuerySet)
```

millerdev commented on tests/test_manager.py:
```diff
@@ -0,0 +1,110 @@
+        self.assertEqual(type(OrderLT40AsManager.objects.all()), LT40QuerySet)
+
+    def test_cte_queryset_correct_manager_n_from_queryset(self):
+        self.assertTrue(isinstance(
```
Could use `self.assertIsInstance(...)` here (and possibly other places too). The error message is more helpful if it fails.

millerdev reviewed (approved):
Thanks. Looks good. Couple minor suggestions.

pgammans commented:
Change ans suggestions applied.

millerdev reviewed (approved):
Thanks for working through all of the feedback!

One nitpicky point from a reviewers perspective: I think you've been using the rebase and force push workflow to keep the history clean, which is nice. However, it makes it difficult to do follow-up reviews because it's impossible to see what changed on the PR. With this workflow the reviewer must start from scratch and re-review the entire contribution from beginning to end, which is tedious. I prefer force push not be used once I've done a review and left comments requiring follow-up.
